### ISSUE TITLE
"KL" ITS doesn't start right

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#1647]
As reported by @eswenson1.  Running the dumped @ ITS from DSKDMP makes it start the Salvager right away.

The dumping procedure isn't right.  It uses $L to load ITS BIN, then $$L to merge in SALV BIN.  However, this gets the starting address from ITS BIN.  It should be 774000 to enter DDT.

I believe adding a `..starta/ ddt` will fix this.

**Comment by eswenson1** [comment]
That doesn't appear to work (tried ..starta/ddt and ..starta/jrst ddt.
```
 DSKDMP
ddt
$L ITS BIN

$$L  SALV BIN

..STARTA/   JRST 122406   JRST DDT
$Y CITS

$G
```

That hangs.

And if reboot into DSKDMP, and just type "cits", I get an infinite stream of BUGHLT messages:
```
 DSKDMP
cits
UUO IN NULL JOB
BUGHALT.  FIND A WIZARD OR CONSIDER TAKING A CRASH DUMP.
THE SYSTEM HAS CRASHED AND CANNOT BE REVIVED WITHOUT EXPERT ATTENTION.
IF YOU CAN'T FIND HELP, RELOAD THE SYSTEM.
YOU ARE NOW IN DDT.
UUO IN NULL JOB
BUGHALT.  FIND A WIZARD OR CONSIDER TAKING A CRASH DUMP.
THE SYSTEM HAS CRASHED AND CANNOT BE REVIVED WITHOUT EXPERT ATTENTION.
IF YOU CAN'T FIND HELP, RELOAD THE SYSTEM.
YOU ARE NOW IN DDT.
UUO IN NULL JOB
BUGHALT.  FIND A WIZARD OR CONSIDER TAKING A CRASH DUMP.
THE SYSTEM HAS CRASHED AND CANNOT BE REVIVED WITHOUT EXPERT ATTENTION.
IF YOU CAN'T FIND HELP, RELOAD THE SYSTEM.
...
```

**Comment by eswenson1** [comment]
Looks like the original starting address is "more correct" than JRST DDT:

```
 DSKDMP
ddt
DDT=774000
$L ITS BIN

..STARTA/   JRST 122406
122406/   SKIPN 200000
122407/   JSR 105642
122410/   JSR 200001
122411/   JFCL 17,122412
122412/   JRST 122413
122413/   JFCL 1,123172
122414/   MOVE 15,117743
122415/   HRRZ 1,..STAR+2
122416/   SUBI 1,100
122417/   LSH 1,-12
122420/   MOVEM 1,74642
122421/   MOVEI 10,2
122422/   SKIPE 41244(10)
122423/   JRST 122436
122424/   MOVE 16,41217(10)
122425/   SUBI 1,1
122426/   MOVE 2,123173
122427/   MOVEM 2,74650(1)
122430/   DPB 10,123174
122431/   SOJG 16,122425
122432/   MOVE 2,1
122433/   LSH 2,12
122434/   HRLI 2,640000
122435/   MOVEM 2,41167(10)
122436/   SOJGE 10,122422
122437/   MOVEM 1,123125
122440/   MOVEI 2,-200(1)
```
Note that this code corresponds to the BEG label in ITS.  And the code there jumps to SALV.  So that looks right. I don't think JRST DDT for a starting address is correct.

**Comment by eswenson1** [comment]
Typing BEG/ gives us the above, with symbols:
```
BEG/'   SKIPN 200000
BEG+1/   JSR BUGAWF
BEG+2/   JSR 200001
GO/   JFCL U,GO+1
GO+1/   JRST GO+2
GO+2/   JFCL A,T00AZ1+12
GO+3/   MOVE P,SYSPDP
GO+4/   HRRZ A,..STAR+2
GO+5/   SUBI A,PATB
GO+6/   LSH A,-12
GO+7/   MOVEM A,FDDTPG
GO+10/   MOVEI Q,B
BEGTT1/   SKIPE QACT(Q)
BEGTT1+1/   JRST BEGTT3
BEGTT1+2/   MOVE T,NTBL(Q)
BEGTT2/   SUBI A,A
BEGTT2+1/   MOVE B,T00AZ1+13
BEGTT2+2/   MOVEM B,MEMBLT(A)
BEGTT2+3/   DPB Q,T00AZ1+14
BEGTT2+4/   SOJG T,BEGTT2
BEGTT2+5/   MOVE B,A
BEGTT2+6/   LSH B,12
BEGTT2+7/   HRLI B,640000
BEGTT2+10/   MOVEM B,QTUTO(Q)
BEGTT3/   SOJGE Q,BEGTT1
BEGTT3+1/   MOVEM A,FTUTPG
BEGTT3+2/   MOVEI B,-200(A)
BEGTT3+3/   LSH B,-1
BEGTT3+4/   ADD B,T00AZ1+15
BEGTT3+5/   TRNE A,A
BEGTT3+6/   IBP B
```

**Comment by larsbrinkhoff** [comment]
Yes, the starting address in ITS BIN is BEG, which is as expected.

Exec DDT only has single-stepping on a KS10.

The normal way to start ITS on a KL10, and I temporarily forgot this, is to $L it in DDT and then $G to start it.

*However*, I would expect that a binary dumped out with $Y from DDT should be able to be run from DSKDMP.  Maybe there's a bug in DDT, or maybe my expectation is wrong.

**Comment by eswenson1** [comment]
Geez, how did anyone debug MC's failures?  Can't believe there was no single stepping!

**Comment by larsbrinkhoff** [comment]
I checked @ ITS dumped from DSKDMP on a KA10 and KS10.  The starting address is BEG.  So clearly my understanding of how ITS starts is wrong.

When you type "its" in DSKDMP, it will load .; @ ITS and then go to DDT.  Since the starting address is BEG, why doesn't ITS start right away?

**Comment by eswenson1** [comment]
Well, what makes you think it isn't doing that?  The first thing ITS does it invoke SALVAGER.  So the message from SALVAGER makes sense.  However, after it invokes SALVAGER, it is supposed to return to ITS and bring up the system.  I wonder if we're having the same problem with RH10 that I was seeing -- something about disk status not being "good" and SALVAGER hanging.

**Comment by rcornwell** [comment]
It is possible that ITS is tickling a part of the RH10 that I did not emulate. The docs on the RP04 and how it sets registers are not very clear. I would suggest enabling debugging of CONI/CONO/DATAIO and seeing how ITS is talking to RH10.

**Comment by larsbrinkhoff** [comment]
Because right after I type Enter, I'm in DDT.  Not in ITS calling SALV.  Only when I type $G, ITS is started and SALVAGER runs.

**Comment by larsbrinkhoff** [comment]
The above applies to starting KA10 or KS10 ITS from DSKDMP.

**Comment by eswenson1** [comment]
That's not what I'm seeing, Lars.

When I start up DSKDMP, and type "ITS<CR>", the "ITS" line is overwritten and my screen looks like this:
```
KA-10 simulator V4.0-0 Current        git commit id: 2ff87d85

 DSKDMP
SALVAGER.314

```
I'm not at DDT.  I'm hung.  This is with a plain pdp10-kl build (no modifications).

**Comment by larsbrinkhoff** [comment]
Right, that's what happens with pdp10-kl.  But this is not what I consider the normal case.

The normal case are the pdp10-ka, simh, or klh10 builds.  These are the KA10 and KS10 versions of ITS.  pdp10-kl is a KA/KL10 hybrid.

This is really tangential to this issue, sorry about that.  But it points to some fault in my understanding.

**Comment by eswenson1** [comment]
Ok.  I'm wondering if there isn't an issue as rcornwell suggested in the RH10 handling.  Since typing DDT to DSKDMP starts DDT fine, I would expect typing ITS to DSKDMP would start it as well -- as long as ITS had a DDT merged in with it.  

Do you see any evidence in documentation (or ITS mailing lists) that you had to use:
```
DDT
$L ITS
$G
```
on MC (KL version)?

**Comment by larsbrinkhoff** [comment]
It appears DDT's $Y only dumps core below the DDT symbol table.  So DDT isn't in the dumped core image.  This totally explains why setting ..STARTA/ DDT doesn't work.

**Comment by larsbrinkhoff** [comment]
Yes, how to start ITS on MC is documented: https://github.com/PDP-10/its/issues/744#issuecomment-374502241

Note, DDT isn't started from DSKDMP, since DSKDMP isn't involved at all.  So it's just `$L ITS` and then `$G`

**Comment by eswenson1** [comment]
Well, what I said (above) certainly doesn't work:
```
KA-10 simulator V4.0-0 Current        git commit id: 2ff87d85

 DSKDMP
ddt
$L ITS BIN

$$L  SALV BIN

$$L @ DDT

$Y AITS

$G
SALVAGER.314


KL ITS 1648 IN OPERATION


KA-10 simulator V4.0-0 Current        git commit id: 2ff87d85

 DSKDMP
SALVAGER.314

```
In other words, even if we merge in DDT (which, yes, I realize should already be there), I get the same behavior -- ITS invokes SALVAGER, it prints out SALVAGER, and then we hang.

**Comment by eswenson1** [comment]
> Yes, how to start ITS on MC is documented: [#744 (comment)](https://github.com/PDP-10/its/issues/744#issuecomment-374502241)
> 
> Note, DDT isn't started from DSKDMP, since DSKDMP isn't involved at all. So it's just `$L ITS` and then `$G`

Oh, good point.  I've been loading DDT first each time. I guess it isn't necessary.

**Comment by larsbrinkhoff** [comment]
Even though you type `$$L @ DDT`, it will not be included when $Y writes the core image.

**Comment by larsbrinkhoff** [comment]
Yes, with the pdp10-kl build we still rely on DSKDMP being there to start DDT.  This is stop gap measure until DDT can be started without DSKDMP.

**Comment by larsbrinkhoff** [comment]
I think the conclusion is the answer to "KL" ITS doesn't start right from DSKDMP is that it should not be started from DSKDMP.  Or if you do, the expected behaviour is to launch ITS and then SALV right away.

**Comment by eswenson1** [comment]
My reading of #744 suggests that MC was never started from DSKDMP, only from NTSDDT and only by doing a $L ITS followed by a $G.  And NTSDDT was started by "J NTSDDT" from KLDCP.

But none of that explains why we can't start ITS from DSKDMP in the same way on our pdp10-kl (KA/RH10) system, as on a KA/RP10 system.

**Comment by larsbrinkhoff** [comment]
I agree about how MC was started.

I also agree that it's not explained why the pdp10-kl ITS behaves differently.  But there is a difference in the pdp10-kl build: it dumps @ ITS from DDT, not DSKDMP.  This means @ ITS doesn't include DDT and DSKDMP like the others do. Now, if you start DSKDMP before ITS it will be resident in memory anyway.
